module4-authentication-and-security/securing-backend-app/README.md

```diff
@@ -0,0 +1,79 @@
+
+In this module we will go over how we secure our app, and how to make sure that neither intended, or not intended security breach happens.
+
+Follow me in the following journey to understand why we need security.
+
+## Wishlist inventory app
+
+You've created an amazing app to store a shopping wishlist for your users. You have end points that retrieves the list, add an item to the list, update an item in the list, and delete an item in the list.
+
+I liked the app and started using it to create 2 lists for the things that I love, computers and kitchenware, and started adding items to both.
+
+In your database you have 2 tables that stores these items as follows:
+
+### Table 1: lists
+
+| id  | name           |
+| --- | -------------- |
+| 1   | Computer Parts |
+| 2   | Kitchenware    |
+
+### Table 2: items
+
+| id  | name                 | list_id | url                   | price | bought |
+| --- | -------------------- | ------- | --------------------- | ----- | ------ |
+| 1   | NVidia RTX 3070      | 1       | https://amzn.com/313  | $785  | false  |
+| 2   | Philips Blender      | 2       | https://url.com/313   | $90   | false  |
+| 3   | AMD Ryzen 9          | 1       | https://amzn.com/585  | $599  | true   |
+| 4   | Tefal Iceforce knife | 2       | https://amzn.com/1311 | $20   | false  |
+
+Now, suppose another user comes to your app, how will your app decides what to show them? When they land on the homepage, they would see my items because the code can't decide what to show. This is the first security risk, the app needs somehow to relate the data to a specific user, so it only query data that belongs to the requesting user.
+
+That's why almost everywhere online, you need to register or login to an account in order to use an application. So tha data can be related to you and you only.
+
+## Let's add some security
+
+To secure our wishlist above, we can add a new column to the lists table to attach it to an identity (a user). We can name that column `user_email` for example, and use this query to get the user wishlist:
+
+```SQL
+SELECT `id`, `name`, `user_email` from `lists`
+INNER JOIN `items`
+ON `items`.`list_id` = `lists`.`id`
+WHERE `lists`.`user_email` = "test@domain.com"
+```
+
+Thus, when a user comes to your app, you can ask them first about their email address, and then query the database to get thier lists. But that is also insufficient. A malicious user can peep on other users wishlists if he/she knew thier emails, or he/she check different email combinations to see if it lands a hit, aka, **Brute Force**.
+
+To make this a bit harder to brute force, you can create a token system. Tokens are random strings of charcters, usually wih arbitrary length, that is intended to be hard to brute force (guess). A token can look something like this: `1FsahNjuio31OHsOallsn1gHH7`
```

Comment:
is this token usage actually a thing? can you point me to a resource? i've never seen someone store things like this. i don't think you would ever change the user_email column to token

i see on line 54 that you mention its "very acceptable" but i don't think i've ever seen a site that asks the user to save a token like this. i think its misleading; this is fine for the purposes of example to show that you don't want to let people brute force something by querying by email but if you're going to do this, you should write on line 38 in very clear text that "this is simply a basic example, but not something we do in practice"

Reply:
additionally if we were to write a site that uses a token system like this that maps emails to tokens, you most likely would have a Users table with the columns `email`, `token`, etc and then your lists app would have a foreign key to that table

overall my concern here is that, i understand this is being used as an introductory example, but also students really take examples to heart and i would not want any student to think that this schema is viable

Reply:
Yes, magic links are actually token-only access providers. I know the architecture requires more, but if we eliminated the clutter, in the end, it is simply a resource guarded by a token. The same goes for collaboration live sessions etc, drive sharing links, AWS s3 URLs.. etc.

What's intended here is not to implement a complete schema, but if you follow a TDD pattern where you incrementally increase the requirements, you would start off with something basic like this. Then when you need to track names and emails, you move into a separate table, etc.

I kinda understand your point though about simplicity, but my approach for the lesson is only to expose what makes the architecture secure enough.

Reply:
i don't think the examples you gave can be considered authentication, actually none of them are guarded by a token

in the example written in the module, you are saying that a list has an owner, but that owner can be replaced by a token, and only the person who knows the token can then go in and query the list

in the other examples you gave, the resource has an ID that happens to be a token, but there is actually almost always an actual underlying owner (maybe not for all collaboration live sessions, but in replit for example there is an "owner" still). 

authentication means that you identify as some entity or user. drive sharing links, aws s3 urls, etc are not tokens that authenticate anyone. they are just tokens, and they happen to identify the thing stored on s3 and the thing stored on drive. there is no authentication involved. those resources are not "guarded" by a token, for example for s3 the ID is a token, no one is using it for security, otherwise that would be known as "security through obscurity" ("oh you can't reach my thing because you don't know the ID"). i think you are confusing an ID token with an auth token

if you google token-based authentication you'll see that almost all results are referring to when you have an existing user already and you can use a token instead of logging in as that user, but that's definitely not the type of auth i'd start off to explain.

finally, arguably you should start off by explaining the most simple and familiar type of authentication...which for everyone is going to be username/password. if i was explaining how to control access to something, i'd start by talking about ACLs, which again is 1. something everyone is familiar with ("i give this person permission to view this") and 2. in terms of teaching security, also the most common way to start, because of (1)

this whole section was confusing even to me. i recommend completely scratching this list/access token example

if you want to include token-based authentication, here would be some examples that students would understand, and it should be placed after username/passwords
* oauth (google/fb login) gives you an oauth token iirc, that authenticates you as a user
* for github you can generate application-specific passwords. these are also like tokens, they are replacement passwords that you can use to login as the same user, but perhaps temporary in scope (but again students probably dont need to know about this)